### ISSUE TITLE
ImportError: cannot import name '_unicodefun' from 'click'

### ISSUE THREAD

**Issue by lucasrodes** [owid/data-utils-py#10]
## System details
- System: 5.10.93-1-MANJARO
- Python: 3.9.5, 3.8.12 (tested in both)

## Experiment

I am trying to install the project in development mode, but I am encountering some errors:

### 1. Run make test

First thing I do is running `make test`, which runs successfully

```
~ make test
poetry install
Creating virtualenv owid-datautils in /home/user/repos/data-utils-py/.venv
Installing dependencies from lock file

Package operations: 54 installs, 0 updates, 0 removals

  • Installing six (1.16.0)
  • Installing asttokens (2.0.5)
  • Installing executing (0.8.3)
  • Installing jmespath (0.10.0)
  • Installing parso (0.8.3)
  • Installing ptyprocess (0.7.0)
  • Installing pure-eval (0.2.2)
  • Installing python-dateutil (2.8.2)
  • Installing traitlets (5.1.1)
  • Installing urllib3 (1.26.8)
  • Installing wcwidth (0.2.5)
  • Installing backcall (0.2.0)
  • Installing botocore (1.24.16)
  • Installing decorator (5.1.1)
  • Installing jedi (0.18.1)
  • Installing matplotlib-inline (0.1.3)
  • Installing pexpect (4.8.0)
  • Installing pickleshare (0.7.5)
  • Installing prompt-toolkit (3.0.28)
  • Installing pygments (2.11.2)
  • Installing pyparsing (3.0.7)
  • Installing stack-data (0.2.0)
  • Installing typing-extensions (4.1.1)
  • Installing attrs (21.4.0)
  • Installing botocore-stubs (1.24.16)
  • Installing click (8.0.4)
  • Installing iniconfig (1.1.1)
  • Installing ipython (8.1.1)
  • Installing mccabe (0.6.1)
  • Installing mypy-boto3-s3 (1.21.23.post1)
  • Installing mypy-extensions (0.4.3)
  • Installing numpy (1.22.3)
  • Installing packaging (21.3)
  • Installing pathspec (0.9.0)
  • Installing platformdirs (2.5.1)
  • Installing pluggy (1.0.0)
  • Installing py (1.11.0)
  • Installing pycodestyle (2.7.0)
  • Installing pyflakes (2.3.1)
  • Installing pytz (2021.3)
  • Installing s3transfer (0.5.2)
  • Installing toml (0.10.2)
  • Installing tomli (1.2.3)
  • Installing black (21.12b0)
  • Installing boto3 (1.21.16)
  • Installing boto3-stubs (1.21.25)
  • Installing colorama (0.4.4)
  • Installing flake8 (3.9.2)
  • Installing ipdb (0.13.9)
  • Installing mypy (0.910)
  • Installing pandas (1.4.1)
  • Installing pytest (6.2.5)
  • Installing structlog (21.5.0)
  • Installing watchdog (2.1.6)

Installing the current project: owid-datautils (0.1.0)
touch .venv
==> Checking formatting
All done! ✨ 🍰 ✨
6 files would be left unchanged.
==> Linting
==> Checking types
poetry run mypy --strict -p owid -p tests
Success: no issues found in 7 source files
==> Running unit tests
=========================================================== test session starts ============================================================
platform linux -- Python 3.9.5, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/lucas/repos/data-utils-py
collected 4 items                                                                                                                          

tests/test_owid_cache.py ..                                                                                                          [ 50%]
tests/test_s3.py ..                                                                                                                  [100%]

============================================================ 4 passed in 0.30s =============================================================
```

### 2. Run make watch
Running `make watch` returns an error, as it is missing library `pyyaml`.

```
~ make watch
==> Watching for changes and re-running tests
poetry run watchmedo shell-command -c 'clear; make check-formatting lint check-typing coverage' --recursive --drop .
Traceback (most recent call last):
  File "/home/user/repos/data-utils-py/.venv/bin/watchmedo", line 5, in <module>
    from watchdog.watchmedo import main
  File "/home/user/repos/data-utils-py/.venv/lib/python3.9/site-packages/watchdog/watchmedo.py", line 34, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'
make: *** [Makefile:23: watch] Error 1
```

### 3. Solving above's error
To solve previous error, I run `poetry update`

```
~ poetry update
Updating dependencies
Resolving dependencies... (5.4s)

Writing lock file

Package operations: 0 installs, 11 updates, 0 removals

  • Updating jmespath (0.10.0 -> 1.0.0)
  • Updating urllib3 (1.26.8 -> 1.26.9)
  • Updating botocore (1.24.16 -> 1.24.28)
  • Updating botocore-stubs (1.24.16 -> 1.24.28)
  • Updating click (8.0.4 -> 8.1.0)
  • Updating ipython (8.1.1 -> 8.2.0)
  • Updating mypy-boto3-s3 (1.21.23.post1 -> 1.21.27.post1)
  • Updating pytz (2021.3 -> 2022.1)
  • Updating boto3 (1.21.16 -> 1.21.28)
  • Updating boto3-stubs (1.21.25 -> 1.21.28)
```

### 4. Finally, re-run make watch

This leads to the following error, which seems to be related to https://github.com/dask/distributed/issues/6013

```
~  make test
poetry install
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: owid-datautils (0.1.0)
touch .venv
==> Checking formatting
Traceback (most recent call last):
  File "/home/user/repos/data-utils-py/.venv/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/user/repos/data-utils-py/.venv/lib/python3.9/site-packages/black/__init__.py", line 1372, in patched_main
    patch_click()
  File "/home/user/repos/data-utils-py/.venv/lib/python3.9/site-packages/black/__init__.py", line 1358, in patch_click
    from click import _unicodefun
ImportError: cannot import name '_unicodefun' from 'click' (/home/user/repos/data-utils-py/.venv/lib/python3.9/site-packages/click/__init__.py)
make: *** [default.mk:27: check-formatting-default] Error 1
```

**Comment by lucasrodes** [comment]
I have changed the version of `black` in [pyproject.toml](https://github.com/owid/data-utils-py/blob/main/pyproject.toml#L24) to `black = "^22.3", ran `poetry update` to reflect the changes in [poetry.lock](https://github.com/owid/data-utils-py/blob/main/poetry.lock), then ran `make watch` and got:

```
==> Checking formatting
All done! ✨ 🍰 ✨
6 files would be left unchanged.
==> Linting
==> Checking types
poetry run mypy --strict -p owid -p tests
Success: no issues found in 7 source files
==> Unit testing with coverage
poetry run pytest --cov=owid --cov-report=term-missing tests
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov=owid --cov-report=term-missing
  inifile: None
  rootdir: /home/user/repos/data-utils-py

```

**Comment by Marigold** [comment]
@lucasrodes nice work, thanks for fixing that. I've tested your setup locally and it works for me now too.

> As per this [source](https://stackoverflow.com/a/26590668/5056599), I added pytest-cov = "^3.0.0" to [pyproject.toml](https://github.com/owid/data-utils-py/blob/main/pyproject.toml#L24)
You can just run `poetry add pytest-cov` and it'll be added to `pyproject.toml` automatically (and you don't have to run `poetry update` then).

> and I think it ran, but it is quite slow

Which part is slow? Wasn't it slow just the first time when everything got compiled?

**Comment by lucasrodes** [comment]
Fixed in #11
